File: modules/coupon-crawler-service/src/main/java/com/thanh0x/coursedeal/tools/ExtractorDebugMain.java
package com.thanh0x.coursedeal.tools;

import com.thanh0x.coursedeal.crawler_runner.CourseDataExtractor;
import com.thanh0x.coursedeal.model.coupon.CouponCourseData;

/**
 * Debug entrypoint to run {@link CourseDataExtractor} for a single URL.
 *
 * Usage (example):
 * UDEMY_DEBUG_DUMP=true UDEMY_DEBUG_URL="<couponUrl>" \
 *   ./gradlew :modules:coupon-crawler-service:debugExtractor
 */
public class ExtractorDebugMain {

    public static void main(String[] args) {
        String url = extractUrlArg(args);
        if (url == null || url.isBlank()) {
            url = System.getenv("UDEMY_DEBUG_URL");
        }
        if (url == null || url.isBlank()) {
            url = System.getProperty("udemy.debugUrl");
        }
        if (url == null || url.isBlank()) {
            throw new IllegalArgumentException("Missing URL. Provide --url=<couponUrl> or set UDEMY_DEBUG_URL.");
        }

        CourseDataExtractor extractor = new CourseDataExtractor(url.trim());
        CouponCourseData result = extractor.getFullCouponCodeData();

        System.out.println("==== Extractor Debug Result ====");
        System.out.println("inputUrl=" + url);
        System.out.println("courseId=" + extractor.getCourseId());
        System.out.println("result=" + (result == null ? "null" : result));
    }

    private static String extractUrlArg(String[] args) {
        if (args == null) return null;
        for (String arg : args) {
            if (arg == null) continue;
            if (arg.startsWith("--url=")) return arg.substring("--url=".length());
            if (arg.equals("--url")) continue; // next arg handled in a second pass below
        }

        // fallback: look for pattern "--url" <value>
        for (int i = 0; i < args.length - 1; i++) {
            if ("--url".equals(args[i])) return args[i + 1];
        }
        return null;
    }
}

File: modules/coupon-crawler-service/src/main/kotlin/com/thanh0x/coursedeal/tools/ExtractorDebugMain.kt
package com.thanh0x.coursedeal.tools

import com.thanh0x.coursedeal.crawlerrunner.CourseDataExtractor

/**
 * Debug entrypoint to run [CourseDataExtractor] for a single URL.
 *
 * Usage (example):
 * UDEMY_DEBUG_DUMP=true UDEMY_DEBUG_URL="<couponUrl>" \
 *   ./gradlew :modules:coupon-crawler-service:debugExtractor
 */
object ExtractorDebugMain {
    @JvmStatic
    fun main(args: Array<String>) {
        val resolvedUrl =
            (extractUrlArg(args) ?: System.getenv("UDEMY_DEBUG_URL") ?: System.getProperty("udemy.debugUrl"))
                ?.trim()
                ?.takeIf { it.isNotBlank() }

        requireNotNull(resolvedUrl) { "Missing URL. Provide --url=<couponUrl> or set UDEMY_DEBUG_URL." }

        val extractor = CourseDataExtractor(resolvedUrl)
        val result = extractor.getFullCouponCodeData()

        println("==== Extractor Debug Result ====")
        println("inputUrl=$resolvedUrl")
        println("courseId=${extractor.courseId}")
        println("result=${result ?: "null"}")
    }

    private fun extractUrlArg(args: Array<String>?): String? {
        if (args == null) return null

        val inlineUrl = args.firstOrNull { it.startsWith("--url=") }?.substringAfter("--url=")
        val separateUrlIndex = args.indexOf("--url")
        val separateUrl = separateUrlIndex.takeIf { it in 0 until args.lastIndex }?.let { args[it + 1] }

        return inlineUrl ?: separateUrl
    }
}
